language.txt

```diff
@@ -288,11 +314,22 @@ SEQUENCE `aabbcc`
 #    c. If the tag is BOOLEAN and the body is valid, encode as TRUE or FALSE.
 #       Otherwise encode as a hex literal.
 #
-#    d. If the tag is BIT STRING and the body is non-empty, encode the first
-#       byte as a separate hex literal. If this value is non-zero, encode the
-#       remainder as a raw byte string. Otherwise, apply step g to encode the
-#       body. This is to account for X.509 incorrectly using BIT STRING
-#       instead of OCTET STRING for SubjectPublicKeyInfo and signatures.
+#    d. If the tag is BIT STRING, try the following until one succeeds:
+#       
+#       i.   If the body is exactly `00`, encode as b``.
+#
+#       ii.  If the first byte is zero, and the body happens to be valid BER,
+#            encode as `00` followed by the results of step g. This, and step
+#            iv, account for X.509 incorrectly using BIT STRING instead of
+#            OCTET STRING for SubjectPublicKeyInfo and signatures.
+#
+#       iii. If the first byte is at most 7, and there are between one and four
+#            bytes after it, encode it as a bit string literal.
+#
+#       iv.  If the first byte is at most 7, encode that byte as a single-byte
+#            byte string, and then encode the rest of it as a byte string.
+#
+#       v.   Give up and encode it as a single byte string.
```

Comment:
I think we can fold (i) into (iii). I also wonder if it's better to talk about whether it's a valid bit string.

```
d. If the tag is a BIT STRING:

   i.   If the body is a valid bit string, contains a whole number of bytes, and
        may be parsed as a series of BER elements with no trailing data, encode
        as `00` followed by recursing into the body as in step g. This accounts
        for X.509 incorrectly using BIT STRING instead of OCTET STRING for
        SubjectPublicKeyInfo and signatures.

   ii.  If the body is a valid bit string with at most 32 bits, encode as a bit
        string literal. If any padding bits are non-zero, they are encoded
        explicitly.

   iii. If the body is a valid bit string with more than 32 bits, encode as a
        pair of hex literals, containing the initial byte and the data bytes.

   iv.  Otherwise, the body is not a valid bit string. Encode as a single
        hex literal.
```

Reply:
That is way better.